LDAP-support for group-owners

joshmoore opened the pull request:
A new property `new_user_group_owner` has been
added which is used by `NewUserGroupOwnerBean`
implementations to query for groups that some
user should be considered an owner of. If the
sync_on_login property is set and the user is
loses the *owner* attribute, then the flag in
OMERO will also be removed.

Currently, only supported for `:query:` based
groups.

#### Testing ####

This began largely as a test of https://github.com/joshmoore/apacheds-docker to set up general LDAP testing. Turns out adding support was easy enough in that context that I figured I'd open a PR.

* **0. Test that the LDAP connection works for you and get a feel for what data (users & groups) is present:**
```
$ ldapsearch -h $SERVER -p 10389 -x -D uid=admin,ou=system -W -b dc=example,dc=com
Enter LDAP Password:
```
To explain the values:
 * `-x` is the type of authentication being used
 * `-D`is the "bind DN", i.e. the identity that you are using to login.
 * `-W` says to ask for a password. A password can be passed on the command-line with `-w`.
 * `-b` is the base of the search (like `omero.ldap.base`). It prevents the query from showing information unrelated to the mythical company "example.com".


* **1. Connect to such an apacheds instance with a configuration of the form:**
```
omero.ldap.config=true
omero.ldap.base=dc=example,dc=com
omero.ldap.group_filter=(objectClass=groupOfUniqueNames)
omero.ldap.group_mapping=name=cn
omero.ldap.new_user_group=:query:(uniquemember=@{dn})
omero.ldap.new_user_group_owner=(owner=@{dn})
omero.ldap.password=$PASSWORD
omero.ldap.sync_on_login=true
omero.ldap.urls=ldap://$SERVER:10389
omero.ldap.user_filter=(objectClass=person)
omero.ldap.user_mapping=omeName=uid,firstName=givenName,lastName=sn,email=mail
omero.ldap.username=uid=admin,ou=system
```


* **2. Using the [./user](https://github.com/joshmoore/apacheds-docker/blob/0cf5e7eeb15a298f784d22c59576fcdf16641e94/user) script which you must download from apacheds-docker, set up users:**
```
$ export LDAPHOST=$SERVERHERE
$ export LDAPPORT=10389
./user ldapuser1 group   # Boot straps the group "test"
./user ldapuser1 add     # Adds the user
./user ldapuser2 add     # Another user
./user ldapuser2 owner # Make an owner
```


* **3. Login as the above users and check their group membership. E.g.:**
```
$ bin/omero -C login ldapuser1@localhost -w ldapuser1
$ bin/omero user info
```


* **4. Change the state of a user and then re-test. E.g.:**
```
$ ./user ldapuser2 member  # Make a non-owner (regular member)
$ bin/omero -C login ldapuser2@localhost -w ldapuser2
$ bin/omero user info
```


* **5. Repeat 3 & 4 as desired, including removing a user from the `test` group and showing that they can no longer log in.**
```
$ ./user ldapuser1 out
$ bin/omero -C login ldapuser1@localhost -w ldapuser1 # Fails
$ ./user ldapuser1 in
$ bin/omero -C login ldapuser1@localhost -w ldapuser1 # Passes
```

pwalczysko commented:
Step 3 of the workflow:
  - having on trying to login as the newly created ldap user:

```
[petr@LS25165 ~/Downloads/OMERO.server-5.1.1-229-8bab540-ice35-b497]$ bin/omero -C login ldapuser1@localhost -w ldapuser1
Internal error. Please contact your administrator:
 Wrapped Exception: (org.springframework.ldap.CommunicationException):
localhost:389; nested exception is javax.naming.CommunicationException: localhost:389 [Root exception is java.net.ConnectException: Connection refused]
```

aleksandra-tarkowska commented:
I've been playing with that PR and our lifesci server using the following config:
```
omero.ldap.base=ou=lifesci,o=dundee
omero.ldap.config=true
omero.ldap.group_filter=(&(objectClass=groupOfUniqueNames)(cn=CLS-Image-Repository))
omero.ldap.group_mapping=name=cn
omero.ldap.new_user_group=:query:(uniqueMember=@{dn)
omero.ldap.new_user_group_owner=(manager=@{dn})
omero.ldap.password=***
omero.ldap.sync_on_login=true
omero.ldap.urls=ldap://ldap.lifesci:389
omero.ldap.user_filter=(objectClass=inetorgperson)
omero.ldap.username=***
```

I already had user (myself) in my database. When I tried to log in I was removed from current group and left unsigned.

```
2015-05-11 16:34:02,661 WARN  [ome.services.sessions.SessionManagerImpl] (l.Server-2) Exception while running executeDefaultGroup
ome.conditions.ValidationException: The user 52 has no default group set.
	at ome.logic.AdminImpl.getDefaultGroup(AdminImpl.java:875) ~[server.jar:na]
	at ome.services.sessions.SessionManagerImpl._getDefaultGroup(SessionManagerImpl.java:1301) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.validateSessionInputs(SessionManagerImpl.java:824) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.access$000(SessionManagerImpl.java:90) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl$2.doWork(SessionManagerImpl.java:358) [server.jar:na]
	at sun.reflect.GeneratedMethodAccessor285.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:566) [server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154) [server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111) [org.springframework.orm.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108) [org.springframework.transaction.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249) [server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121) [server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at com.sun.proxy.$Proxy74.doWork(Unknown Source) [na:na]
	at ome.services.util.Executor$Impl.execute(Executor.java:447) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.createSession(SessionManagerImpl.java:352) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.createFromRequest(SessionManagerImpl.java:276) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.createWithAgent(SessionManagerImpl.java:296) [server.jar:na]
	at ome.services.blitz.fire.SessionManagerI.create(SessionManagerI.java:183) [blitz.jar:na]
	at Glacier2._SessionManagerDisp.___create(_SessionManagerDisp.java:114) [ice-glacier2.jar:na]
	at Glacier2._SessionManagerDisp.__dispatch(_SessionManagerDisp.java:148) [ice-glacier2.jar:na]
	at IceInternal.Incoming.invoke(Incoming.java:222) [ice.jar:na]
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482) [ice.jar:na]
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258) [ice.jar:na]
	at Ice.ConnectionI.message(ConnectionI.java:1213) [ice.jar:na]
	at IceInternal.ThreadPool.run(ThreadPool.java:321) [ice.jar:na]
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12) [ice.jar:na]
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693) [ice.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_31]
```

group wasn;t created. Did I do something wrong?

joshmoore commented:
> group wasn;t created. Did I do something wrong?

@aleksandra-tarkowska: what groups were you in, are you, and were you *expecting* to be in?

aleksandra-tarkowska commented:
I had local group 'data'. According to group filter I'd expect sync_on_login should create only one matching filter. Am I wrong? Is sync_on_login in sync with group filters as well?

aleksandra-tarkowska commented:
My expectation was that following config will create group CLS-Image-Repository add Kenny as an owner and me as a member. Is that incorrect?

joshmoore commented:
Is he also a member of that group? And "he" attempted a login?

aleksandra-tarkowska commented:
kenny is a manager, he own the group.
I tried to add him 
```
Mac-Book-Pro:openmicroscopy ola$ dist/bin/omero ldap create khgillen
Previously logged in to localhost:4064 as ola
Server: [localhost:4064]
Username: [ola]
Password:
Created session 61a48520-6590-4d6b-bc5c-d93e44784382 (ola@localhost:4064). Idle timeout: 60.0 min. Current group: read-only
No group found for: cn=khgillen,ou=edir,ou=people,ou=lifesci,o=dundee
Mac-Book-Pro:openmicroscopy ola$ dist/bin/omero ldap create jrswedlow
Using session 61a48520-6590-4d6b-bc5c-d93e44784382 (ola@localhost:4064). Idle timeout: 60.0 min. Current group: read-only
No group found for: cn=jrswedlow,ou=edir,ou=people,ou=lifesci,o=dundee
Mac-Book-Pro:openmicroscopy ola$ 
```
no success

aleksandra-tarkowska commented:
The group look like cn=CLS-Image-Repository,...o=dundee

```
cn=CLS-Image-Repository

displayName=CLS-Image-Repository

manager=cn=khgillen,...o=dundee

memberUid:
jrswedlow
rkferguson
atarkowska
cblackburn

uniqueMember
cn=atarkowska,...o=dundee
cn=cblackburn,...o=dundee
cn=jrswedlow,...o=dundee
cn=rkferguson,...o=dundee
```

joshmoore commented:
At the moment, you will also need to make khgillen a member of the group.

aleksandra-tarkowska commented:
I am sorry, I cannot get that working. First, I only can log in as user who's account was created in the past if ```omero.ldap.sync_on_login=false``` and I manually created group matching ldap group ```CLS-Image-Repository```

@kennethgillen added himself to the group

cn=CLS-Image-Repository
manager=cn=khgillen,ou=edir,ou=people,ou=lifesci,o=dundee

memberUid:
atarkowska
rkferguson
cblackburn
jrswedlow
khgillen

uniqueMember:
cn=atarkowska,ou=e...sci,o=dundee
cn=cblackburn,ou=e...sci,o=dundee
cn=jrswedlow,ou=e...sci,o=dundee
cn=rkferguson,ou=e...sci,o=dundee
cn=khgillen,ou=e...sci,o=dundee

and when I tried to add him I am getting
```
$ dist/bin/omero ldap create khgillen
Previous session expired for ola on localhost:4064
...
Created session b5792752-a7ed-4057-8c90-d8c6af287f01 (ola@localhost:4064). Idle timeout: 60.0 min. Current group: read-only
No group found for: cn=khgillen,ou=e...sci,o=dundee
```

I also asked Kenny to log in, User wasn;t created at all.

```
2015-05-14 13:49:30,366 INFO  [  ome.security.auth.LdapPasswordProvider] (l.Server-8) Default choice on create user: khgillen (ome.conditions.ValidationException: No group found for: cn=khgillen,ou=e...sci,o=dundee)
```

Second ```omero.ldap.sync_on_login=true``` doesn't work at all.
I cannot even log in as existing user

Is there anything wrong with my config?

joshmoore commented:
@aleksandra-tarkowska, on the ldap2 instance, you'll need to change your group filter to:
```
omero.ldap.group_filter=(&(objectClass=groupOfUniqueNames)(cn=CLS-Image-Repository))
```
i.e. s/uid/cn/

aleksandra-tarkowska commented:
Thanks @joshmoore for pointing missed ```}``` now config seems to be ok but I am getting

```
2015-05-15 11:29:27,951 DEBUG [ ome.security.auth.QueryNewUserGroupBean] (l.Server-7) Running query: (&(&(objectClass=groupOfUniqueNames)(cn=CLS-Image-Repository))(uniqueMember=uid=jmoore,ou=Users,dc=example,dc=com))
2015-05-15 11:29:27,978 DEBUG [ ome.security.auth.QueryNewUserGroupBean] (l.Server-7) Running query: (&(&(objectClass=groupOfUniqueNames)(cn=CLS-Image-Repository))(uniqueMember=uid=jmoore,ou=Users,dc=example,dc=com)(owner=uid=jmoore,ou=Users,dc=example,dc=com))
2015-05-15 11:29:27,990 INFO  [       ome.security.basic.CurrentDetails] (l.Server-7) Adding log:INSERT,class ome.model.meta.Experimenter,1453
2015-05-15 11:29:28,002 INFO  [       ome.security.basic.CurrentDetails] (l.Server-7) Adding log:INSERT,class ome.model.meta.GroupExperimenterMap,2405
2015-05-15 11:29:28,010 INFO  [       ome.security.basic.CurrentDetails] (l.Server-7) Adding log:INSERT,class ome.model.meta.GroupExperimenterMap,2406
2015-05-15 11:29:28,013 INFO  [                 org.perf4j.TimingLogger] (l.Server-7) start[1431685767884] time[128] tag[omero.call.exception]
2015-05-15 11:29:28,014 WARN  [        ome.services.util.ServiceHandler] (l.Server-7) Unknown exception thrown.

java.lang.NullPointerException: null
	at ome.security.auth.SimpleRoleProvider.setGroupOwner(SimpleRoleProvider.java:162) ~[server.jar:na]
	at ome.logic.LdapImpl.createUser(LdapImpl.java:510) ~[server.jar:na]
	at ome.logic.LdapImpl.createUser(LdapImpl.java:453) ~[server.jar:na]
	at ome.security.auth.LdapPasswordProvider.checkPassword(LdapPasswordProvider.java:113) ~[server.jar:na]
	at ome.security.auth.PasswordProviders.checkPassword(PasswordProviders.java:52) ~[server.jar:na]
	at ome.logic.AdminImpl.checkPassword(AdminImpl.java:1197) ~[server.jar:na]
	at ome.services.sessions.SessionManagerImpl$9.doWork(SessionManagerImpl.java:1050) ~[server.jar:na]
	at sun.reflect.GeneratedMethodAccessor288.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:566) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111) ~[org.springframework.orm.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108) ~[org.springframework.transaction.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at com.sun.proxy.$Proxy74.doWork(Unknown Source) [na:na]
	at ome.services.util.Executor$Impl.execute(Executor.java:447) [server.jar:na]
	at ome.services.util.Executor$Impl.execute(Executor.java:391) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executeCheckPasswordRW(SessionManagerImpl.java:1045) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executeCheckPassword(SessionManagerImpl.java:1017) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executePasswordCheck(SessionManagerImpl.java:993) [server.jar:na]
	at ome.services.blitz.fire.PermissionsVerifierI.checkPermissions(PermissionsVerifierI.java:135) [blitz.jar:na]
	at Glacier2._PermissionsVerifierDisp.___checkPermissions(_PermissionsVerifierDisp.java:114) [ice-glacier2.jar:na]
	at Glacier2._PermissionsVerifierDisp.__dispatch(_PermissionsVerifierDisp.java:149) [ice-glacier2.jar:na]
	at IceInternal.Incoming.invoke(Incoming.java:222) [ice.jar:na]
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482) [ice.jar:na]
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258) [ice.jar:na]
	at Ice.ConnectionI.message(ConnectionI.java:1213) [ice.jar:na]
	at IceInternal.ThreadPool.run(ThreadPool.java:321) [ice.jar:na]
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12) [ice.jar:na]
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693) [ice.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_31]
2015-05-15 11:29:28,015 INFO  [        ome.services.util.ServiceHandler] (l.Server-7)  Excp:	java.lang.NullPointerException
2015-05-15 11:29:28,016 ERROR [     o.s.blitz.fire.PermissionsVerifierI] (l.Server-7) Exception thrown while checking password for:jmoore
ome.conditions.InternalException:  Wrapped Exception: (java.lang.NullPointerException):
null
	at ome.security.auth.SimpleRoleProvider.setGroupOwner(SimpleRoleProvider.java:162) ~[server.jar:na]
	at ome.logic.LdapImpl.createUser(LdapImpl.java:510) ~[server.jar:na]
	at ome.logic.LdapImpl.createUser(LdapImpl.java:453) ~[server.jar:na]
	at ome.security.auth.LdapPasswordProvider.checkPassword(LdapPasswordProvider.java:113) ~[server.jar:na]
	at ome.security.auth.PasswordProviders.checkPassword(PasswordProviders.java:52) ~[server.jar:na]
	at ome.logic.AdminImpl.checkPassword(AdminImpl.java:1197) ~[server.jar:na]
	at ome.services.sessions.SessionManagerImpl$9.doWork(SessionManagerImpl.java:1050) ~[server.jar:na]
	at sun.reflect.GeneratedMethodAccessor288.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:566) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.security.basic.EventHandler.invoke(EventHandler.java:154) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111) ~[org.springframework.orm.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108) ~[org.springframework.transaction.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at com.sun.proxy.$Proxy74.doWork(Unknown Source) ~[na:na]
	at ome.services.util.Executor$Impl.execute(Executor.java:447) ~[server.jar:na]
	at ome.services.util.Executor$Impl.execute(Executor.java:391) ~[server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executeCheckPasswordRW(SessionManagerImpl.java:1045) ~[server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executeCheckPassword(SessionManagerImpl.java:1017) ~[server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executePasswordCheck(SessionManagerImpl.java:993) ~[server.jar:na]
	at ome.services.blitz.fire.PermissionsVerifierI.checkPermissions(PermissionsVerifierI.java:135) ~[blitz.jar:na]
	at Glacier2._PermissionsVerifierDisp.___checkPermissions(_PermissionsVerifierDisp.java:114) [ice-glacier2.jar:na]
	at Glacier2._PermissionsVerifierDisp.__dispatch(_PermissionsVerifierDisp.java:149) [ice-glacier2.jar:na]
	at IceInternal.Incoming.invoke(Incoming.java:222) [ice.jar:na]
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482) [ice.jar:na]
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258) [ice.jar:na]
	at Ice.ConnectionI.message(ConnectionI.java:1213) [ice.jar:na]
	at IceInternal.ThreadPool.run(ThreadPool.java:321) [ice.jar:na]
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12) [ice.jar:na]
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693) [ice.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_31]
```

aleksandra-tarkowska commented:
Found small issue with ```sync_on_login```. Basically if group was created by hand in the past and its name match group on your ldap server, when you turn on ```sync_on_login``` this group is not recognized as ldap group. It may be debatable how that should behave as before 5.1 we didn;t have ldap field on ExperimenterGroup and that may be confusing why user remain in that group.

aleksandra-tarkowska commented:
@joshmoore thanks for quick fix of NPE. With the ```omero.ldap.group_filter=(&(objectClass=groupOfUniqueNames)(cn=OME))``` I managed to log in as owner and all looks ok

![firefoxscreensnapz041](https://cloud.githubusercontent.com/assets/1065155/7651961/6a5ff89e-fafe-11e4-8765-1b679f950c12.png)

Now I removed the group filter to use only ```omero.ldap.group_filter=(objectClass=groupOfUniqueNames)```. It is expected that when I log in as an owner ```sync_on_login``` will pick up 2 other groups.

I cannot log in as any user
This time I am getting another error:
```
2015-05-15 12:34:29,921 DEBUG [  ome.security.auth.LdapPasswordProvider] (l.Server-4) lookupLdap(1502)=uid=jmoore,ou=Users,dc=example,dc=com
2015-05-15 12:34:29,929 DEBUG [  ome.security.auth.LdapPasswordProvider] (l.Server-4) findDN(jmoore)=uid=jmoore,ou=Users,dc=example,dc=com
2015-05-15 12:34:29,943 DEBUG [ ome.security.auth.QueryNewUserGroupBean] (l.Server-4) Running query: (&(objectClass=groupOfUniqueNames)(uniqueMember=uid=jmoore,ou=Users,dc=example,dc=com))
2015-05-15 12:34:29,954 DEBUG [ ome.security.auth.QueryNewUserGroupBean] (l.Server-4) Running query: (&(objectClass=groupOfUniqueNames)(uniqueMember=uid=jmoore,ou=Users,dc=example,dc=com)(owner=uid=jmoore,ou=Users,dc=example,dc=com))
2015-05-15 12:34:29,966 INFO  [                      ome.logic.LdapImpl] (l.Server-4) Adding groups for jmoore: [1253, 1303]
2015-05-15 12:34:29,992 INFO  [                 org.perf4j.TimingLogger] (l.Server-4) start[1431689669865] time[126] tag[omero.call.exception]
2015-05-15 12:34:29,998 WARN  [        ome.services.util.ServiceHandler] (l.Server-4) Unknown exception thrown.

org.springframework.jdbc.UncategorizedSQLException: PreparedStatementCallback; uncategorized SQLException for SQL [select ome_nextval(?,?)]; SQL state [25P02]; error code [0]; ERROR: current transaction is aborted, commands ignored until end of transaction block; nested exception is org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands ignored until end of transaction block
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:83) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:80) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:602) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:636) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:665) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:673) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.queryForObject(JdbcTemplate.java:728) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.queryForObject(JdbcTemplate.java:744) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.queryForLong(JdbcTemplate.java:765) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.simple.SimpleJdbcTemplate.queryForLong(SimpleJdbcTemplate.java:128) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at ome.util.actions.PostgresSqlAction.nextValue(PostgresSqlAction.java:218) ~[model-psql.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_31]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_31]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.util.SqlAction$LoggingSqlAction.invoke(SqlAction.java:82) ~[model-psql.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at com.sun.proxy.$Proxy3.nextValue(Unknown Source) ~[na:na]
	at ome.util.TableIdGenerator.generate(TableIdGenerator.java:53) ~[model-psql.jar:na]
	at org.hibernate.event.def.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:122) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.DefaultSaveOrUpdateEventListener.saveWithGeneratedOrRequestedId(DefaultSaveOrUpdateEventListener.java:210) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.DefaultSaveOrUpdateEventListener.entityIsTransient(DefaultSaveOrUpdateEventListener.java:195) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.DefaultSaveOrUpdateEventListener.performSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:117) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.DefaultSaveOrUpdateEventListener.onSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:93) ~[hibernate-core.jar:na]
	at org.hibernate.impl.SessionImpl.fireSaveOrUpdate(SessionImpl.java:679) ~[hibernate-core.jar:na]
	at org.hibernate.impl.SessionImpl.saveOrUpdate(SessionImpl.java:671) ~[hibernate-core.jar:na]
	at org.hibernate.engine.CascadingAction$5.cascade(CascadingAction.java:252) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascadeToOne(Cascade.java:392) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascadeAssociation(Cascade.java:335) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascadeProperty(Cascade.java:204) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascadeCollectionElements(Cascade.java:425) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascadeCollection(Cascade.java:362) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascadeAssociation(Cascade.java:338) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascadeProperty(Cascade.java:204) ~[hibernate-core.jar:na]
	at org.hibernate.engine.Cascade.cascade(Cascade.java:161) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.AbstractFlushingEventListener.cascadeOnFlush(AbstractFlushingEventListener.java:154) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.AbstractFlushingEventListener.prepareEntityFlushes(AbstractFlushingEventListener.java:145) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:88) ~[hibernate-core.jar:na]
	at org.hibernate.event.def.DefaultDirtyCheckEventListener.onDirtyCheck(DefaultDirtyCheckEventListener.java:54) ~[hibernate-core.jar:na]
	at org.hibernate.impl.SessionImpl.isDirty(SessionImpl.java:1194) ~[hibernate-core.jar:na]
	at sun.reflect.GeneratedMethodAccessor278.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307) ~[org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.tools.hibernate.SessionFactory.invoke(SessionFactory.java:104) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at com.sun.proxy.$Proxy75.isDirty(Unknown Source) ~[na:na]
	at ome.security.basic.EventHandler.invoke(EventHandler.java:178) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.orm.hibernate3.HibernateInterceptor.invoke(HibernateInterceptor.java:111) ~[org.springframework.orm.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108) ~[org.springframework.transaction.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121) ~[server.jar:na]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202) [org.springframework.aop.jar:3.0.1.RELEASE-A]
	at com.sun.proxy.$Proxy74.doWork(Unknown Source) [na:na]
	at ome.services.util.Executor$Impl.execute(Executor.java:447) [server.jar:na]
	at ome.services.util.Executor$Impl.execute(Executor.java:391) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executeCheckPasswordRO(SessionManagerImpl.java:1024) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executeCheckPassword(SessionManagerImpl.java:1015) [server.jar:na]
	at ome.services.sessions.SessionManagerImpl.executePasswordCheck(SessionManagerImpl.java:993) [server.jar:na]
	at ome.services.blitz.fire.PermissionsVerifierI.checkPermissions(PermissionsVerifierI.java:135) [blitz.jar:na]
	at Glacier2._PermissionsVerifierDisp.___checkPermissions(_PermissionsVerifierDisp.java:114) [ice-glacier2.jar:na]
	at Glacier2._PermissionsVerifierDisp.__dispatch(_PermissionsVerifierDisp.java:149) [ice-glacier2.jar:na]
	at IceInternal.Incoming.invoke(Incoming.java:222) [ice.jar:na]
	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482) [ice.jar:na]
	at Ice.ConnectionI.dispatch(ConnectionI.java:1258) [ice.jar:na]
	at Ice.ConnectionI.message(ConnectionI.java:1213) [ice.jar:na]
	at IceInternal.ThreadPool.run(ThreadPool.java:321) [ice.jar:na]
	at IceInternal.ThreadPool.access$300(ThreadPool.java:12) [ice.jar:na]
	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693) [ice.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_31]
Caused by: org.postgresql.util.PSQLException: ERROR: current transaction is aborted, commands ignored until end of transaction block
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2270) ~[postgresql.jar:9.4]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1998) ~[postgresql.jar:9.4]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:255) ~[postgresql.jar:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:570) ~[postgresql.jar:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:420) ~[postgresql.jar:9.4]
	at org.postgresql.jdbc2.AbstractJdbc2Statement.executeQuery(AbstractJdbc2Statement.java:305) ~[postgresql.jar:9.4]
	at sun.reflect.GeneratedMethodAccessor105.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_31]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_31]
	at bitronix.tm.resource.jdbc.BaseProxyHandlerClass.invoke(BaseProxyHandlerClass.java:64) ~[btm.jar:2.1.3]
	at com.sun.proxy.$Proxy8.executeQuery(Unknown Source) ~[na:na]
	at org.springframework.jdbc.core.JdbcTemplate$1.doInPreparedStatement(JdbcTemplate.java:643) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:586) ~[org.springframework.jdbc.jar:3.0.1.RELEASE-A]
	... 79 common frames omitted
```

joshmoore commented:
So for anyone watching along, in the above there were three different classes of errors:
 * NPE; fixed by https://github.com/joshmoore/openmicroscopy/commit/4c6a60cf726655cf98aadb07ef53feaa7b26c27c
 * Configuration errors (`${dn)` rather than `${dn})`); fixed in @aleksandra-tarkowska's local config
 * `ERROR: current transaction is aborted` which is only happening in @aleksandra-tarkowska's local install. I could not reproduce with her DB dump and her corrected config.

aleksandra-tarkowska commented:
if anyone knows why I am getting:

```
uncategorized SQLException for SQL [select ome_nextval(?,?)]; SQL state [25P02]; error code [0];
```
and how to fix it, you know how to find me :)

mtbc commented:
I believe this is caused by a *previous* SQL command failing.

aleksandra-tarkowska commented:
see https://trac.openmicroscopy.org/ome/ticket/12887

joshmoore commented:
To simplify further testing, @sbesson will try to get this configured in the daily merge build. (with my thanks!)

sbesson commented:
Tested using trout/integration (and a Docker LDAP) via OMERO.web:
- `./user ldapuser1 add && ./user ldapuser1 in` - login adds a new LDAP user
- `./user ldapuser1 owner` - login changes ldapuser1 to  group owner
- `./user ldapuser1 out`  - cannot login
- `./user ldapuser1 in && ./user ldapuser1 member` - can login again and user is demoted as group member

Added a second user and a second group holding both users, all combinations of membership/ownership in LDAP led to the expected result in OMERO after re-logging in.

Do we need to test  other LDAP configurations than the one described in this PR?
Also I haven't tested the `sync_on_login` boolean but I assume it suffers from the same basic limitations (LDAP vs non LDAP groups)

joshmoore commented:
`sync_on_login` was fixed as of https://github.com/openmicroscopy/openmicroscopy/pull/3761.

sbesson commented:
Tested the following workflow using the configuration in the PR description:
- in LDAP tree, create LDAP user in LDAP group
- log into to OMERO - both user and group are created
- add the user to another non-LDAP group - log out
-  in LDAP tree, remove the LDAP user from the LDAP group 
- log in to OMERO - user is only member of the non-LDAP group - log out
-  in LDAP tree, re-adds the LDAP user in the LDAP group and make him owner of the group
- log in to OMERO - user is now member of the non-LDAP group and owner of the LDAP group

All works as expected. Merging.
